BloomFilter.md

```diff
@@ -181,6 +182,9 @@ struct ColumnMetaData {
 
 ```
 
+The Bloom filter data is stored right after the page indexes, and the file layout looks like:
```

Comment:
Why is this stored after page indexes? The page index data is going to be used more often and is smaller, so I think it should be just before the header and bloom filters should be before the indexes.

Why is this layout required? Was there performance testing done adding the bloom filters here instead of at the start of row groups?

Reply:
There is no specific reason why the bloom filter is stored after page indexes. IIUC, both page indexes and bloom filter are random reads, will they impact each other? 

Regarding store bloom filter at the footer, that is because when using HDFS seeking in a parquet file may require remote reads which consuming remote I/O and bandwidth and we can't reading parquet footer in any case. So that we can save some  I/O when storing bloom filters in the footer.

Reply:
Page indexes aren't necessarily going to require random reads. Some readers guess the length of the footer and read, for example, 1MB of data at the end of the file to avoid a backward seek. If that includes the entire footer and page indexes, then it doesn't require a second backward seek. And it will become more likely that the indexes are in that 1MB chunk when we update Parquet to use truncated lower and upper bounds.

Anyway, the point is that the bloom filters shouldn't be placed between the page indexes and the footer.

> because when using HDFS seeking in a parquet file may require remote reads which consuming remote I/O and bandwidth

I don't think this is true. Usually, the footer is read remotely.

A task is created for an HDFS block and assigned where that block is local. When that task runs, it will read the footer, which usually probably remote, find which row groups are in the assigned block, and then read the matching row groups that are probably local. Since bloom filter data is large, I think you would end up with more remote reads in HDFS by locating them near the footer.

Keeping bloom filters in memory to be placed near the footer is also more expensive for memory consumption because Parquet can't just write out the filter data and reuse that memory.

I'm not sure what the right choice is here.

@jbapple, @julienledem, and @zivanfi: any opinions about where bloom filters should be located?

Reply:
Understood your point, thanks for the elaboration.

Reply:
@rdblue Thank you for asking. I'm not yet well informed enough about the details to have an opinion.

Reply:
Just updated the layout according to Ryan's comments, I can update the layout if there is any update after discussion.

Reply:
@chenjunjiedada, to get this done, let's allow both as an option. So bloom filters may be located either between row groups (e.g., bloom filter col 1, bloom filter col 3, column chunk 1, column chunk 2, column chunk 3) OR may be located before the page indexes after all row groups.

In both cases, bloom filters should be grouped by row group with data for each column in the same order as the file schema. That will reduce seeks for row group data when processing a subset of the row groups, which is the common access pattern from processing splits.

Reply:
Agreed, maybe we can provide an option in writer to select the location.

Reply:
@rdblue, format updated. Please take a look.